File: 1912.rb
def max?(a,b,c)
  if a > b && a > c
    return a
  elsif b > a && b > c
    return b
  else
    return c
  end
end

def sequal_max(nums)
  max = 0
  sum = 0
  nums.each do |num|
    sum += num
    max = sum if sum > max
  end
  return max
end

def largest?(nums)
  l = nums.length
  if l === 1
    nums[0] 
  elsif l === 2
    max?(nums[0], nums[1], (nums[0]+nums[1]))
  elsif l % 2
    half = l/2-1
    left = nums[0..half].reverse
    right = nums[(half+1)..(l-1)]
    return max?((sequal_max(left) + sequal_max(right)), largest?(left), largest?(right))
  else
    half = l/2
    left = nums[0..(half-1)].reverse
    right = nums[(half+1)..(l-1)]
    return max?((sequal_max(left) + nums[half] + sequal_max(right)), largest?(left), largest?(right))
  end
end

N = gets.chomp.to_i
nums = gets.chomp.split(' ').map{|e| e.to_i }

p largest?(nums)

File: 1912.cpp
#include <cstdio>

int n, max, sequalSum;
int nums[100000];

int largest(int nums[])
{
  int start = -1;
  max = -1000;
  for(int i=0; i < n; i++)
  {
    if(nums[i] > 0){
      start = i;
      break;
    }
    if(nums[i] > max){
      max = nums[i]; 
    }
  }
  if(start == -1){ return max; }
  sequalSum = nums[start];
  max = nums[start];
  for(int i=(start + 1); i < n ; i++)
  {
    sequalSum += nums[i];
    if(nums[i] < 0 && sequalSum >= 0){
      continue;
    } else if(sequalSum < 0){
      sequalSum = 0;
    } else if (nums[i] >= 0 && sequalSum >= max){
      max = sequalSum;
    } else {
      continue;
    }
  }
  return max;
}

int main()
{
  scanf("%d",&n);
  for(int i=0; i < n; i++)
    scanf(" %d",&nums[i]);
  
  printf("%d",largest(nums));
  return 0;
}

10
10 -4 3 1 5 6 -35 12 21 -1